### DISCUSSION TITLE
Confusing error "needs to have a constructor with 0 args or only optional args"

### DISCUSSION THREAD

**Discussion by thomaslevesque** [LuckyPennySoftware/AutoMapper#3609] [Ideas]
### Source/destination types

```csharp
// Source
record Foo(string x, int y);
// Destination
record FooDto(string x, int z); // Note the z instead of y
```

### Mapping configuration

```csharp
cfg.CreateMap<Foo, FooDto>();
```

### Expected behavior

Should give an error saying that it can't resolve a value for constructor argument `z`, or something similar.

### Actual behavior

Gives the following error:

> ArgumentException: FooDto needs to have a constructor with 0 args or only optional args. (Parameter 'type')

This is confusing, because FooDto doesn't *actually* need to have a constructor with 0 args or only optional args (it works fine with `record FooDto(string x, int y)`). The problem is really that it can't resolve the value for `z`, so the error should say that.

I started encountering the error quite frequently since I started using records. If there's a mismatch between properties of the source and destination types, I get this error. Now I know what to look for, but the first time I saw it I was really confused and it took me a while to understand what the problem was.

### Steps to reproduce

```csharp
mapper.Map<FooDto>(foo);
```

**Comment by lbargaoanu** [comment]
That seems difficult in the general case. But you're welcome to try a PR.

**Reply by thomaslevesque** [reply]
> That seems difficult in the general case.

Yes, I was looking at the code right now, it doesn't seem straightforward. Still, I think it would be a big improvement, given the number of issues that mention this error. If it's too complex, maybe just change the message to something like this?

> FooDto needs to have a constructor with 0 args, only optional args, or only args whose value can be resolved from the source type.

It's not perfect, but it would be less misleading.

(Nitpicking: it should probably say "parameters" rather than "args". Parameters are what you declare in the constructor, arguments are what you pass when you call it)

**Reply by lbargaoanu** [reply]
A lot of people have no idea that constructor mapping exists :) So I'm not convinced this is an improvement. I see your point though.

**Reply by lbargaoanu** [reply]
Anyway, you should be calling `AssertConfigurationIsValid`. That one says `No available constructor'. Not that helpful perhaps, but at least closer to the truth.

**Reply by thomaslevesque** [reply]
> A lot of people have no idea that constructor mapping exists :)

Sorry, but that's not a very compelling argument... A lot of people do know about it, and with records it's going to be more and more common to use constructor mapping.

> So I'm not convinced this is an improvement.

Given that the current message is actually wrong (since the type doesn't actually need a parameterless constructor), I think it can't hurt to improve it.

> Anyway, you should be calling AssertConfigurationIsValid. That one says `No available constructor'. Not that helpful perhaps, but at least closer to the truth.

It also mentions the unmapped property, which *is* helpful.

> Unmapped members were found. Review the types and members below.
> Add a custom mapping expression, ignore, add a custom resolver, or modify the source/destination type
> For no matching constructor, add a no-arg ctor, add optional arguments, or map all of the constructor parameters
> \==================================
> Foo -> FooDto (Destination member list)
> UserQuery+Foo -> UserQuery+FooDto (Destination member list)
> 
> **Unmapped properties:**
> **z**
> No available constructor.

I guess that's a good thing to do anyway.

BTW, what is the recommended way to validate configuration when using `AutoMapper.Extensions.Microsoft.DependencyInjection`? There are a few [issues](https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection/issues?q=is%3Aissue+AssertConfigurationIsValid+is%3Aclosed) about it, but they're either closed without an answer, or suggest using the static Mapper class, which is no longer recommended.

**Reply by bartdebever** [reply]
Would like to add to this as I've just ran into this error trying to map a class to a record. Doing this makes the error even more unclear than from record to record. Only due to the existance of this thread I understand and can fix the issue.
Think this should have a better error detailing which property is not mapped.

> A lot of people have no idea that constructor mapping exists :)

This is true yet shouldn't be true, it's a powerful feature especially with records on the rise and can only see it needing to be used more and more in the future.
